### DISCUSSION TITLE
ControlNet units order: is that important

### DISCUSSION THREAD

**Discussion by IPv6** [Mikubill/sd-webui-controlnet#1932] [General]
Always wanted to know - is there any meaningful difference for units order in multi-controlnet inference setup?
I mean when same unit (for example pose, ref, softedge, normal) placed at start of unit sequence - do unit gets more "priority"? Or when placed at end - are there any dependency on results of the work of previous units?

Or this is commutative thing and order is not relevant?

**Comment by ljleb** [comment]
You can test this manually by pinning the seed. I was able to swap 2 different controlnet models and get exactly the same image in each case 2-3 months ago, not sure if anything changed that makes this impossible now.

**Reply by IPv6** [reply]
@ljleb actually i see the difference with same seed and reversed order - but it is really subtle. Like some zones darker and other small fluctuations. So not sure is this something that can be used or needs to be generally ignored
![grid-0054](https://github.com/Mikubill/sd-webui-controlnet/assets/40495/1a089d0b-7182-43bc-879a-2881dd38caa9)
![grid-0055](https://github.com/Mikubill/sd-webui-controlnet/assets/40495/47af3e5c-ef77-4e24-9542-415077c1075e)

**Reply by huchenlei** [reply]
according to my test, shuffle and reference are not fully deterministic with fixed seed/subseed even when used individually. Make sure your test do not use these 2 control types.

**Reply by ljleb** [reply]
These differences may be caused by the digital precision of the tensors (I assume fp16).
Are the images as different when using `--no-half`?

**Reply by ljleb** [reply]
For the record, I get very similar differences when swapping text condition vectors along the axis with size 77 (so preserving the tensors of 768). Sometimes even some colours flip, but it's usually very local changes.

**Comment by IPv6** [comment]
> according to my test, shuffle and reference are not fully deterministic with fixed seed/subseed even when used individually. Make sure your test do not use these 2 control types.

Yes, reference was used in this. So except this the order of other types not relevant? and with ref/shuffle - it is not the order, that may cause changes? Nice insight, thank you.

After some tests seems order is not important (at least that much), but would be cool to know is this native to composing method... and may be changes in composing math could help to enforce some control units over other

To be honest CN weight is not that reliable - for pose/etc values less than 1 frequently produce inconsistent poses (extra legs/hands/etc) instead of pose variances (which is really wanted with "less weight" slider)
